File: Examples/test-suite/go/char_strings_runme.go
package main

import (
	"strconv"
	"swigtests/char_strings"
)

const CPLUSPLUS_MSG = "A message from the deep dark world of C++, where anything is possible."
const OTHERLAND_MSG = "Little message from the safe world."

func main() {
	const count = 1000

	// get functions
	for i := uint(0); i < count; i++ {
		if char_strings.GetCharHeapString() != CPLUSPLUS_MSG {
			panic("Test char get 1 failed, iteration " + strconv.Itoa(int(i)))
		}
		char_strings.DeleteCharHeapString()
	}

	for i := uint(0); i < count; i++ {
		if char_strings.GetConstCharProgramCodeString() != CPLUSPLUS_MSG {
			panic("Test char get 2 failed, iteration " + strconv.Itoa(int(i)))
		}
		char_strings.DeleteCharHeapString()
	}

	for i := uint(0); i < count; i++ {
		if char_strings.GetCharStaticString() != CPLUSPLUS_MSG {
			panic("Test char get 3 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if char_strings.GetCharStaticStringFixed() != CPLUSPLUS_MSG {
			panic("Test char get 4 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if char_strings.GetConstCharStaticStringFixed() != CPLUSPLUS_MSG {
			panic("Test char get 5 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	// set functions
	for i := uint(0); i < count; i++ {
		if !char_strings.SetCharHeapString(OTHERLAND_MSG+strconv.Itoa(int(i)), i) {
			panic("Test char set 1 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if !char_strings.SetCharStaticString(OTHERLAND_MSG+strconv.Itoa(int(i)), i) {
			panic("Test char set 2 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if !char_strings.SetCharArrayStaticString(OTHERLAND_MSG+strconv.Itoa(int(i)), i) {
			panic("Test char set 3 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if !char_strings.SetConstCharHeapString(OTHERLAND_MSG+strconv.Itoa(int(i)), i) {
			panic("Test char set 4 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if !char_strings.SetConstCharStaticString(OTHERLAND_MSG+strconv.Itoa(int(i)), i) {
			panic("Test char set 5 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if !char_strings.SetConstCharArrayStaticString(OTHERLAND_MSG+strconv.Itoa(int(i)), i) {
			panic("Test char set 6 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if !char_strings.SetCharConstStaticString(OTHERLAND_MSG+strconv.Itoa(int(i)), i) {
			panic("Test char set 7 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if !char_strings.SetConstCharConstStaticString(OTHERLAND_MSG+strconv.Itoa(int(i)), i) {
			panic("Test char set 8 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if !char_strings.SetConstCharTypedefString(OTHERLAND_MSG+strconv.Itoa(int(i)), i) {
			panic("Test char set 9 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	// get set function
	for i := uint(0); i < count*10; i++ {
		ping := OTHERLAND_MSG + strconv.Itoa(int(i))
		pong := char_strings.CharPingPong(ping)
		if ping != pong {
			panic("Test PingPong 1 failed.\nExpected:" + ping + "\nReceived:" + pong)
		}
	}

	for i := uint(0); i < count; i++ {
		ping := OTHERLAND_MSG + strconv.Itoa(int(i))
		pong := char_strings.CharArrayPingPong(ping)
		if ping != pong {
			panic("Test PingPong 2 failed.\nExpected:" + ping + "\nReceived:" + pong)
		}
	}

	for i := uint(0); i < count; i++ {
		ping := OTHERLAND_MSG + strconv.Itoa(int(i))
		pong := char_strings.CharArrayDimsPingPong(ping)
		if ping != pong {
			panic("Test PingPong 3 failed.\nExpected:" + ping + "\nReceived:" + pong)
		}
	}

	// variables
	for i := uint(0); i < count; i++ {
		expected := OTHERLAND_MSG + strconv.Itoa(int(i))
		char_strings.SetGlobal_char(expected)
		if char_strings.GetGlobal_char() != expected {
			panic("Test variables 1 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		expected := OTHERLAND_MSG + strconv.Itoa(int(i))
		char_strings.SetGlobal_char_array1(expected)
		if char_strings.GetGlobal_char_array1() != expected {
			panic("Test variables 2 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		expected := OTHERLAND_MSG + strconv.Itoa(int(i))
		char_strings.SetGlobal_char_array2(expected)
		if char_strings.GetGlobal_char_array2() != expected {
			panic("Test variables 3 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if char_strings.GetGlobal_const_char() != CPLUSPLUS_MSG {
			panic("Test variables 4 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if char_strings.GetGlobal_const_char_array1() != CPLUSPLUS_MSG {
			panic("Test variables 5 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if char_strings.GetGlobal_const_char_array2() != CPLUSPLUS_MSG {
			panic("Test variables 6 failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	// char *& tests
	for i := uint(0); i < count; i++ {
		if char_strings.GetCharPointerRef() != CPLUSPLUS_MSG {
			panic("Test char pointer ref get failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if !char_strings.SetCharPointerRef(OTHERLAND_MSG+strconv.Itoa(int(i)), i) {
			panic("Test char pointer ref set failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if char_strings.GetConstCharPointerRef() != CPLUSPLUS_MSG {
			panic("Test const char pointer ref get failed, iteration " + strconv.Itoa(int(i)))
		}
	}

	for i := uint(0); i < count; i++ {
		if !char_strings.SetConstCharPointerRef(OTHERLAND_MSG+strconv.Itoa(int(i)), i) {
			panic("Test const char pointer ref set failed, iteration " + strconv.Itoa(int(i)))
		}
	}
}

File: Examples/test-suite/r/char_strings_runme.R
clargs <- commandArgs(trailing=TRUE)
source(file.path(clargs[1], "unittest.R"))

dyn.load(paste("char_strings", .Platform$dynlib.ext, sep=""))
source("char_strings.R")
cacheMetaData(1)

CPLUSPLUS_MSG = "A message from the deep dark world of C++, where anything is possible."
OTHERLAND_MSG = "Little message from the safe world."

count = 5

# get functions
for (i in 0:(count-1)) {
  unittest(GetCharHeapString(), CPLUSPLUS_MSG)
  DeleteCharHeapString()
}

for (i in 0:(count-1)) {
  unittest(GetConstCharProgramCodeString(), CPLUSPLUS_MSG)
  DeleteCharHeapString()
}

for (i in 0:(count-1)) {
  unittest(GetCharStaticString(), CPLUSPLUS_MSG)
}

for (i in 0:(count-1)) {
  unittest(GetCharStaticStringFixed(), CPLUSPLUS_MSG)
}

for (i in 0:(count-1)) {
  unittest(GetConstCharStaticStringFixed(), CPLUSPLUS_MSG)
}

# set functions
for (i in 0:(count-1)) {
  unittest(SetCharHeapString(paste0(OTHERLAND_MSG, i), i), TRUE)
}

for (i in 0:(count-1)) {
  unittest(SetCharStaticString(paste0(OTHERLAND_MSG, i), i), TRUE)
}

for (i in 0:(count-1)) {
  unittest(SetCharArrayStaticString(paste0(OTHERLAND_MSG, i), i), TRUE)
}

for (i in 0:(count-1)) {
  unittest(SetConstCharHeapString(paste0(OTHERLAND_MSG, i), i), TRUE)
}

for (i in 0:(count-1)) {
  unittest(SetConstCharStaticString(paste0(OTHERLAND_MSG, i), i), TRUE)
}

for (i in 0:(count-1)) {
  unittest(SetConstCharArrayStaticString(paste0(OTHERLAND_MSG, i), i), TRUE)
}

for (i in 0:(count-1)) {
  unittest(SetCharConstStaticString(paste0(OTHERLAND_MSG, i), i), TRUE)
}

for (i in 0:(count-1)) {
  unittest(SetConstCharConstStaticString(paste0(OTHERLAND_MSG, i), i), TRUE)
}

for (i in 0:(count-1)) {
  unittest(SetConstCharTypedefString(paste0(OTHERLAND_MSG, i), i), TRUE)
}

# get set functions (ping pong round-trip)
for (i in 0:(count-1)) {
  ping = paste0(OTHERLAND_MSG, i)
  unittest(CharPingPong(ping), ping)
}

for (i in 0:(count-1)) {
  ping = paste0(OTHERLAND_MSG, i)
  unittest(CharArrayPingPong(ping), ping)
}

for (i in 0:(count-1)) {
  ping = paste0(OTHERLAND_MSG, i)
  unittest(CharArrayDimsPingPong(ping), ping)
}

# variables
for (i in 0:(count-1)) {
  expected = paste0(OTHERLAND_MSG, i)
  global_char_set(expected)
  unittest(global_char_get(), expected)
}

# global_char_array1 / global_char_array2 read-back is omitted: the generated R
# getters return type '_p_char' and the R proxy wraps the (correct) string into
# new("_p_char", ref=...), an undefined S4 class which errors.  The setters run
# without error, but there is no working accessor to verify the value.

for (i in 0:(count-1)) {
  unittest(global_const_char_get(), CPLUSPLUS_MSG)
}

# global_const_char_array1 / global_const_char_array2 read-back is omitted for
# the same '_p_char' proxy limitation as the mutable char array globals above.

# char *& tests
for (i in 0:(count-1)) {
  unittest(GetCharPointerRef(), CPLUSPLUS_MSG)
}

for (i in 0:(count-1)) {
  unittest(SetCharPointerRef(paste0(OTHERLAND_MSG, i), i), TRUE)
}

for (i in 0:(count-1)) {
  unittest(GetConstCharPointerRef(), CPLUSPLUS_MSG)
}

for (i in 0:(count-1)) {
  unittest(SetConstCharPointerRef(paste0(OTHERLAND_MSG, i), i), TRUE)
}
